src/coreclr/jit/helperexpansion.cpp

```diff
@@ -583,18 +588,50 @@ bool Compiler::fgExpandThreadLocalAccessForCall(BasicBlock** pBlock, Statement*
     GenTree* maxThreadStaticBlocksValue =
         gtNewIndir(TYP_INT, maxThreadStaticBlocksRef, GTF_IND_NONFAULTING | GTF_IND_INVARIANT);
 
-    // Create tree for "if (maxThreadStaticBlocks < typeIndex)"
-    GenTree* maxThreadStaticBlocksCond =
-        gtNewOperNode(GT_LT, TYP_INT, maxThreadStaticBlocksValue, gtCloneExpr(typeThreadStaticBlockIndexValue));
-    maxThreadStaticBlocksCond = gtNewOperNode(GT_JTRUE, TYP_VOID, maxThreadStaticBlocksCond);
-
     // Create tree for "threadStaticBlockBase = tls[offsetOfThreadStaticBlocks]"
     GenTree* offsetOfThreadStaticBlocks = gtNewIconNode(offsetOfThreadStaticBlocksVal, TYP_I_IMPL);
     GenTree* threadStaticBlocksRef =
         gtNewOperNode(GT_ADD, TYP_I_IMPL, gtCloneExpr(tlsLclValueUse), offsetOfThreadStaticBlocks);
     GenTree* threadStaticBlocksValue =
         gtNewIndir(TYP_I_IMPL, threadStaticBlocksRef, GTF_IND_NONFAULTING | GTF_IND_INVARIANT);
 
+#else
+    GenTree* tls_get_addr_val = gtNewIconHandleNode(threadStaticBlocksInfo.tlsGetAddrFtnPtr, GTF_ICON_FTN_ADDR);
+    tlsValue                  = gtNewIndCallNode(tls_get_addr_val, TYP_I_IMPL);
+    GenTreeCall* tlsRefCall   = tlsValue->AsCall();
+
+    // This is a syscall indirect call which takes an argument.
+    // Populate and set the ABI apporpriately.
+    GenTree* tlsArg = gtNewIconNode(threadStaticBlocksInfo.descrAddrOfMaxThreadStaticBlock, TYP_I_IMPL);
+    tlsRefCall->gtArgs.InsertAfterThisOrFirst(this, NewCallArg::Primitive(tlsArg));
+
+    CallArg* arg0 = tlsRefCall->gtArgs.GetArgByIndex(0);
+    arg0->AbiInfo = CallArgABIInformation();
+    arg0->AbiInfo.SetRegNum(0, REG_ARG_0);
+
+    tlsRefCall->gtFlags |= GTF_EXCEPT | (tls_get_addr_val->gtFlags & GTF_GLOB_EFFECT);
+#ifdef UNIX_X86_ABI
+    tlsRefCall->gtFlags &= ~GTF_CALL_POP_ARGS;
+#endif // UNIX_X86_ABI
+
+    // Cache the tls value
+    GenTree* tlsValueDef              = gtNewStoreLclVarNode(tlsLclNum, tlsValue);
+    GenTree* maxThreadStaticBlocksRef = gtNewLclVarNode(tlsLclNum);
+    GenTree* maxThreadStaticBlocksValue =
+        gtNewIndir(TYP_INT, gtCloneExpr(maxThreadStaticBlocksRef), GTF_IND_NONFAULTING | GTF_IND_INVARIANT);
+
+    GenTree* threadStaticBlocksRef = gtNewOperNode(GT_ADD, TYP_I_IMPL, gtCloneExpr(maxThreadStaticBlocksRef),
```

Comment:
Note: here, I am not using `descrAddrOfThreadStaticBlocks`, but assuming that the `threadStaticBlocks` array is located just after the `maxThreadStaticBlocks` entry in TLS. If we decide to use `descrAddrOfThreadStaticBlocks`, we will have to do another call to `__tls_get_addr()` to get it's address. I was trying to somehow add a `static_assert` to enforce the declaration order, but its not possible on global variables.

Locally, when I verified the addresses by declaring another variable in between `maxThreadStaticBlocks` and `threadStaticBlocks`, but still saw that they are located just a `pointer size` away from each other.

Reply:
This is very fragile assumption. We should do it properly - doing it properly does not require two `__tls_get_addr()` calls.

Reply:
Fixed with a robust model.